cpp/src_prims/selection/knn.h

```diff
@@ -18,179 +18,598 @@
+ * array in each thread. This is going to cause warp divergence of log(n)
+ * per iteration.
+ * @param unique_labels array of unique labels
+ * @param n_labels number of unique labels
+ * @param target_val the label value to search for in unique_labels
+ */
+template <typename IdxType = int>
+__device__ int label_binary_search(IdxType *unique_labels, IdxType n_labels,
+                                   IdxType target_val) {
+  int out_label_idx = -1;
+
+  int level = 1;
+  int cur_break_idx = round(n_labels / (2.0 * level));
+  while (out_label_idx == -1) {
+    int cur_cached_label = unique_labels[cur_break_idx];
+
+    // If we found our label, terminate
+    if (cur_cached_label == target_val) {
+      return cur_break_idx;
+
+      // check left neighbor
+    } else if (cur_break_idx > 0 &&
+               unique_labels[cur_break_idx - 1] == target_val) {
+      return cur_break_idx - 1;
+
+      // check right neighbor
+    } else if (cur_break_idx < n_labels - 1 &&
+               unique_labels[cur_break_idx + 1] == target_val) {
+      return cur_break_idx + 1;
+
+      // traverse
+    } else {
+      level += 1;
+
+      int subtree = round(n_labels / (2.0 * level));
+      if (target_val < cur_cached_label) {
+        // take left subtree
+        cur_break_idx -= subtree;
       } else {
-        std::stringstream ss;
-        ss << "Input memory for " << ptr
-           << " failed. isDevice?=" << att.devicePointer << ", N=" << sizes[i];
-        std::cout << "Exception: " << ss.str() << std::endl;
+        // take right subtree
+        cur_break_idx += subtree;
       }
     }
   }
+  return -1;
+}
 
-  merge_tables<faiss::CMin<float, IntType>>(
-    long(n), k, n_params, result_D, result_I, all_D, all_I, id_ranges->data());
+template <typename OutType = float>
+__global__ void class_probs_kernel(OutType *out, const int64_t *knn_indices,
+                                   const int *labels, int *unique_labels,
+                                   int n_uniq_labels, size_t n_samples,
+                                   int n_neighbors) {
+  int row = (blockIdx.x * blockDim.x) + threadIdx.x;
+  int i = row * n_neighbors;
 
-  MLCommon::updateDevice(res_D, result_D, k * size_t(n), s);
-  MLCommon::updateDevice(res_I, result_I, k * size_t(n), s);
+  extern __shared__ int label_cache[];
 
-  delete all_D;
-  delete all_I;
+  if (threadIdx.x == 0) {
+    for (int j = 0; j < n_uniq_labels; j++) label_cache[j] = unique_labels[j];
+  }
 
-  delete result_D;
-  delete result_I;
-};
+  __syncthreads();
+
+  if (row >= n_samples) return;
+
+  for (int j = 0; j < n_neighbors; j++) {
+    int64_t neighbor_idx = knn_indices[i + j];
+    int out_label = labels[neighbor_idx];
+
+    // Trading off warp divergence in the outputs so that we don't
+    // need to copy / modify the label memory to do these mappings.
+    // Found a middle-ground between between using shared memory
+    // for the mappings.
+    int out_label_idx =
```

Comment:
I don't understand the reason for not doing a pre-processing stage of creating this label mappings to monotonically increasing numbers. Can you explain, please?

My gut feeling is that you'll be much faster with this mapping, than doing binary search on the label values, especially since this needs to be done inside the `n_neighbors` loop.

Reply:
The reasoning is to save memory. I see two options if we decided to do the pre-processing, unless you are able to think of another option:
1) Modify the user's labels array and change it back when done. This doesn't seem safe, especially if they were to call different algorithms in different streams. 
2) Allocate a completely new array for the label mappings. 

I know the binary search option is not going to be as performant as a constant-time mapping, but I am trying to reach a reasonable trade-off between having to duplicate the labels array completely and a little bit slower execution time. What do you think? Do you think it's more reasonable just to use the memory and duplicate the labels?

Reply:
I have created issue #1398 to address this. After thinking about this more, I do agree that I think copying the labels is the best approach here and I am able to justify it because this algorithm so easy to batch. Thus, if a user is concerned about memory footprint, they can break up their data into batches.